src/OpenFeature/ProviderRepository.cs

```diff
@@ -0,0 +1,291 @@
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using OpenFeature.Constant;
+using OpenFeature.Model;
+
+
+namespace OpenFeature
+{
+    /// <summary>
+    /// This class manages the collection of providers, both default and named, contained by the API.
+    /// </summary>
+    internal class ProviderRepository
+    {
+        private FeatureProvider _defaultProvider = new NoOpFeatureProvider();
+
+        private readonly ConcurrentDictionary<string, FeatureProvider> _featureProviders =
+            new ConcurrentDictionary<string, FeatureProvider>();
+
+        /// The reader/writer locks is not disposed because the singleton instance should never be disposed.
+        ///
+        /// Mutations of the _defaultProvider or _featureProviders are done within this lock even though
+        /// _featureProvider is a concurrent collection. This is for a couple reasons, the first is that
+        /// a provider should only be shutdown if it is not in use, and it could be in use as either a named or
+        /// default provider.
+        ///
+        /// The second is that a concurrent collection doesn't provide any ordering so we could check a provider
+        /// as it was being added or removed such as two concurrent calls to SetProvider replacing multiple instances
+        /// of that provider under different names..
+        private readonly ReaderWriterLockSlim _providersLock = new ReaderWriterLockSlim();
+
+        /// <summary>
+        /// Set the default provider
+        /// </summary>
+        /// <param name="featureProvider">the provider to set as the default, passing null has no effect</param>
+        /// <param name="context">the context to initialize the provider with</param>
+        /// <param name="afterSet">
+        /// <para>
+        /// Called after the provider is set, but before any actions are taken on it.
+        /// </para>
+        /// This can be used for tasks such as registering event handlers. It should be noted that this can be called
+        /// several times for a single provider. For instance registering a provider with multiple names or as the
+        /// default and named provider.
+        /// <para>
+        /// </para>
+        /// </param>
+        /// <param name="afterInitialization">
+        /// called after the provider has initialized successfully, only called if the provider needed initialization
+        /// </param>
+        /// <param name="afterError">
+        /// called if an error happens during the initialization of the provider, only called if the provider needed
+        /// initialization
+        /// </param>
+        /// <param name="afterShutdown">called after a provider is shutdown, can be used to remove event handlers</param>
+        public async Task SetProvider(
+            FeatureProvider featureProvider,
+            EvaluationContext context,
+            Action<FeatureProvider> afterSet = null,
```

Comment:
All these "actions" can be used to facilitate implementing events support. I didn't want to put too much stuff together, so this should be a nice middle ground.